src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -747,7 +746,8 @@ buffer_manager_advance_to_non_empty_buffer (
 				// delete the empty buffer
 				EventPipeBuffer *removed_buffer = ep_buffer_list_get_and_remove_head (buffer_list);
 				EP_ASSERT (current_buffer == removed_buffer);
-				buffer_manager_deallocate_buffer (buffer_manager, removed_buffer);
+				buffer_manager_release_buffer(buffer_manager, ep_buffer_get_size (removed_buffer));
+				ep_rt_buffer_array_append(&buffer_manager->used_buffer_array, removed_buffer);
```

Comment:
I think we'll have to do that if we want to free the memory before releasing the buffer budget.  This is the logical place to move the deletions while still keeping it outside the lock.  It has the added benefit of consolidating the release/free semantics back to one call since it would happen in one step again.

I'll move it here and see how that affects the numbers.